File: ErshovVladislav/Task4/Task4/src/main/java/syncronizedset/LazySet.java
package syncronizedset;

public class LazySet<T> implements ISynchronizedSet<T> {

	private Node<T> head = new Node<>(Integer.MIN_VALUE);
	private Node<T> tail = new Node<>(Integer.MAX_VALUE);

	public LazySet() {
		head.setNext(tail);
	}

	@Override
	public boolean add(T value) {
		int key = value.hashCode();
		while (true) {
			Node<T> previous = head;
			Node<T> current = head.getNext();

			while (current.getKey() < key) {
				previous = current;
				current = current.getNext();
			}

			previous.lock();
			try {
				current.lock();
				try {
					if (validate(previous, current)) {
						if (current.getKey() == key) {
							return false;
						} else {
							Node<T> node = new Node<T>(key, value);
							node.setNext(current);
							previous.setNext(node);

							return true;
						}
					}
				} finally {
					current.unlock();
				}
			} finally {
				previous.unlock();
			}
		}
	}

	@Override
	public boolean remove(T value) {
		int key = value.hashCode();
		while (true) {
			Node<T> previous = head;
			Node<T> current = head.getNext();

			while (current.getKey() < key) {
				previous = current;
				current = current.getNext();
			}

			previous.lock();
			try {
				current.lock();
				try {
					if (validate(previous, current)) {
						if (current.getKey() != key) {
							return false;
						} else {
							current.mark();
							previous.setNext(current.getNext());
							return true;
						}
					}
				} finally {
					current.unlock();
				}
			} finally {
				previous.unlock();
			}
		}
	}

	@Override
	public boolean contains(T value) {
		int key = value.hashCode();
		Node<T> current = head;

		while (current.getKey() < key) {
			current = current.getNext();
		}

		return current.getKey() == key && !current.isMarked();
	}

	@Override
	public int size() {
		Node<T> current = head;
		while (!current.equals(tail)) {
			current.lock();
			current = current.getNext();
		}
		current.lock();

		try {
			int count = 0;
			current = head.getNext();
			while (!current.equals(tail)) {
				count++;
				current = current.getNext();
			}

			return count;
		} finally {
			current = head;
			while (!current.equals(tail)) {
				current.unlock();
				current = current.getNext();
			}
			current.unlock();
		}
	}

	private boolean validate(Node<T> previous, Node<T> current) {
		return !previous.isMarked() && !current.isMarked() && previous.getNext().equals(current);
	}

}

File: FominaViktoriia/DeansOffice/DeansOffice/Sets/LazySet.cs
using DeansOffice.ExamSystemLogic;

namespace DeansOffice.Sets
{
    public class LazySet : IExamSystem
    {
        private readonly Node<(long, long)> _tail = new(int.MaxValue);
        private readonly Node<(long, long)> _head = new(int.MinValue);
        private volatile int count;

        public LazySet()
        {
            _head.Next = _tail;
        }

        private static bool Validate(Node<(long, long)> pred, Node<(long, long)> curr)
        {
            return !pred.Marked && !curr.Marked && pred.Next == curr;
        }

        public void Add(long studentId, long courseId)
        {
            var item = (studentId, courseId);
            int key = item.GetHashCode();
            while (true)
            {
                Node<(long, long)> pred = _head;
                Node<(long, long)> curr = _head.Next;
                while (curr.Key < key)
                {
                    pred = curr;
                    curr = curr.Next;
                }
                pred.Lock.Lock();
                try
                {
                    curr.Lock.Lock();
                    try
                    {
                        if (Validate(pred, curr))
                        {
                            if (curr.Key == key)
                            {
                                return;
                            }
                            else
                            {
                                Node<(long, long)> node = new Node<(long, long)>(item) { Next = curr };
                                pred.Next = node;
                                ++count;
                                return;
                            }
                        }
                    }
                    finally
                    {
                        curr.Lock.Unlock();
                    }
                }
                finally
                {
                    pred.Lock.Unlock();
                }

            }
        }

        public void Remove(long studentId, long courseId)
        {
            var item = (studentId, courseId);
            int key = item.GetHashCode();
            while (true)
            {
                Node<(long, long)> pred = _head;
                Node<(long, long)> curr = _head.Next;
                while (curr.Key < key)
                {
                    pred = curr;
                    curr = curr.Next;
                }
                pred.Lock.Lock();
                try
                {
                    curr.Lock.Lock();
                    try
                    {
                        if (Validate(pred, curr))
                        {
                            if (curr.Key != key)
                            {
                                return;
                            }
                            else
                            {
                                curr.Marked = true;
                                pred.Next = curr.Next;
                                --count;
                                return;
                            }
                        }
                    }
                    finally
                    {
                        curr.Lock.Unlock();
                    }
                }
                finally
                {
                    pred.Lock.Unlock();
                }
            }
        }

        public bool Contains(long studentId, long courseId)
        {
            var item = (studentId, courseId);
            int key = item.GetHashCode();
            Node<(long, long)> curr = _head;
            while (curr.Key < key)
                curr = curr.Next;
            return curr.Key == key && !curr.Marked;
        }

        public int Count => count;
    }
}
